### PR TITLE
feat(server-core): superuser guard

### PR DESCRIPTION
**Pull request by aviel-lanzman**
[issues #1 ](https://github.com/Bokoness/lashon/issues/1)

### CODE REVIEW

**Comment by Bokoness** [comment]
Please update branch with master

File: middleware/superUserMiddleware.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	sess, err := services.GetStore(c)
+
+	if err != nil {
+		return fiber.NewError(fiber.StatusInternalServerError)
+	}
+
+	sess.Set("user", user)
+
+	if err = sess.Save(); err != nil {
+		fmt.Printf("this is the error %s", err)
```
Comment (Bokoness) [inline middleware/superUserMiddleware.go:33]: If you can use our logger instead it will be great (here and in the other guard)

**Review by Bokoness — CHANGES_REQUESTED** [review]

File: middleware/superUserMiddleware.go (hunk 1 above)
Comment (aviel-lanzman) [reply middleware/superUserMiddleware.go:33]: @Bokoness 
like this:
`log.Fatal(fmt.Printf("this is the error %s", err))`
?

File: middleware/authMiddleware.go [hunk 2]
```diff
@@ -2,31 +2,49 @@ package middleware
...
+	}
+
+	sess.Set("user", user)
+
+	if err = sess.Save(); err != nil {
+		log.Fatal(fmt.Printf("this is the error %s", err))
+		return statusCodeError
+	}
+
+	return 0
```
Comment (Bokoness) [inline middleware/authMiddleware.go:26]: Why do you return 0? 
if you want to make sure that there is no error, you should return an error instead, and `nil` if there are no errors.
Then in when you call this method you do something like this:
```go
err := InsertUserToSesstion(c, user) 
if err != nil {
  return fiber.NewError(statusCodeError)
 }
 ```

File: middleware/authMiddleware.go [hunk 3]
```diff
@@ -2,31 +2,49 @@ package middleware
...
 	"fmt"
+	"log"
+
+	"github.com/bokoness/lashon/models"
 	"github.com/bokoness/lashon/services"
 	"github.com/gofiber/fiber/v2"
 )
 
-func AuthMiddleware(c *fiber.Ctx) error {
+func InsertUserToSession(c *fiber.Ctx, user *models.User) int {
```
Comment (Bokoness) [inline middleware/authMiddleware.go:12]: Maybe it is best to move those function into auth service

File: middleware/authMiddleware.go [hunk 4]
```diff
@@ -2,31 +2,49 @@ package middleware
...
+	sess.Set("user", user)
+
+	if err = sess.Save(); err != nil {
+		log.Fatal(fmt.Printf("this is the error %s", err))
+		return statusCodeError
+	}
+
+	return 0
+}
+func GetUserFromCookiesByJWT(c *fiber.Ctx) (*models.User, error) {
```
Comment (Bokoness) [inline middleware/authMiddleware.go:28]: Move to authService

**Review by Bokoness — CHANGES_REQUESTED** [review]

File: middleware/superUserMiddleware.go (hunk 1 above)
Comment (Bokoness) [reply middleware/superUserMiddleware.go]: https://docs.gofiber.io/api/log

Comment (Bokoness) [reply middleware/superUserMiddleware.go]: but not fatal, fatal will shutdown the server

File: server/middleware/auth_middleware.go [hunk 5]
```diff
@@ -1,54 +1,25 @@
...
 func AuthMiddleware(c *fiber.Ctx) error {
-	user, err := GetUserFromCookiesByJWT(c)
+	user, err := services.GetUserFromCookiesByJWT(c)
 	if err != nil || user == nil {
 		return c.SendStatus(fiber.StatusUnauthorized)
 	}
 
 	// insert user to session
-	statusCodeError := InsertUserToSession(c, user)
+	statusCodeError := services.InsertUserToSession(c, user)
```
Comment (Bokoness) [inline server/middleware/auth_middleware.go:16]: As I mentioned before, that service shouldn't know anything about status code, its the responsibility of the controller, so the service should return only error, no statusCode.

File: server/services/auth_service.go [hunk 6]
```diff
@@ -100,3 +102,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
 
 	return user, nil
 }
+func InsertUserToSession(c *fiber.Ctx, user *models.User) int {
+	sess, err := GetStore(c)
```
Comment (Bokoness) [inline server/services/auth_service.go:107]: Please use `GetUserFromSession` function instead

File: server/services/auth_service.go [hunk 7]
```diff
@@ -100,3 +102,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
...
+	statusCodeError := fiber.StatusInternalServerError
```
Comment (Bokoness) [inline server/services/auth_service.go:108]: Don't user `statusCodeError` the service shouldn't know anything about status codes, only controller

File: server/middleware/auth_middleware.go [hunk 8]
```diff
@@ -5,23 +5,21 @@ import (
...
+	user, err := services.GetUserFromCookiesByJWT(c)
+	if err != nil || user == nil {
 		return c.SendStatus(fiber.StatusUnauthorized)
 	}
 
-	user, err := services.GetUserByJWT(cookie)
-
-	if err != nil {
-		return fiber.NewError(fiber.StatusUnauthorized)
+	// insert user to session
```
Comment (Bokoness) [inline server/middleware/auth_middleware.go:15]: No need for this comment, the function name is descriptive enough

File: server/middleware/superUserMiddleware.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	user, err := services.GetUserFromCookiesByJWT(c)
+	if err != nil || user == nil {
+		return err
+	}
+
+	if !user.IsSuperuser {
+		return c.SendStatus(fiber.StatusUnauthorized)
+	}
+
+	// insert user to session
```
Comment (Bokoness) [inline server/middleware/superUserMiddleware.go:18]: remove comment

File: server/middleware/superUserMiddleware.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"github.com/bokoness/lashon/services"
+	"github.com/gofiber/fiber/v2"
+)
+
+func SuperUserMiddleware(c *fiber.Ctx) error {
+	user, err := services.GetUserFromCookiesByJWT(c)
+	if err != nil || user == nil {
+		return err
```
Comment (Bokoness) [inline server/middleware/superUserMiddleware.go:11]:
```suggestion
		return c.SendStatus(fiber.StatusUnauthorized)
```

File: server/models/user.go [hunk 11]
```diff
@@ -1,14 +1,20 @@
...
 
 type User struct {
 	gorm.Model
-	Email    string `json:"email"`
-	Password string `json:"password"`
-	FullName string `json:"fullName" gorm:"column: full_name"`
+	Email    	string 	  `json:"email"`
+	Password    string    `json:"password"`
+	FullName    string    `json:"fullName" gorm:"column: full_name"`
+	IsSuperuser bool      `json:"isSuperuser" gorm:"is_superuser;default:false"`
```
Comment (Bokoness) [inline server/models/user.go:14]:
```suggestion
	IsSuperuser bool      `json:"isSuperuser" gorm:"column:is_superuser;default:false"`
```

File: server/models/user.go [hunk 12]
```diff
@@ -1,14 +1,20 @@
...
+	CreatedAt   time.Time `gorm:"autoCreateTime: true"`
+	UpdatedAt   time.Time `gorm:"autoUpdateTime: true"`
+
```
Comment (Bokoness) [inline server/models/user.go:17]:
```suggestion
```
Remove those, since we are using `gorm.Model` in the struct it will be created automatically

**Review by Bokoness — CHANGES_REQUESTED** [review]

File: server/middleware/superUserMiddleware.go [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package middleware
```
Comment (Bokoness) [inline server/middleware/superUserMiddleware.go:1]: Please rename the flie to `superuser_middleware`

File: server/services/auth_service.go (hunk 6 above)
Comment (aviel-lanzman) [reply server/services/auth_service.go:107]: How am I supposed to take the user out of the session if here I put it in the session.
user not supposed to be inside the session

Comment (Bokoness) [reply server/services/auth_service.go:107]: Sorry my bad

File: server/services/auth_service.go [hunk 14]
```diff
@@ -100,3 +103,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
 
 	return user, nil
 }
+func InsertUserToSession(c *fiber.Ctx, user *models.User) error {
```
Comment (Bokoness) [inline server/services/auth_service.go:106]:
```suggestion

func InsertUserToSession(c *fiber.Ctx, user *models.User) error {
```

File: server/middleware/superuser_middleware.go [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+		return c.SendStatus(fiber.StatusUnauthorized)
+	}
+
+	if !user.IsSuperuser {
+		return c.SendStatus(fiber.StatusUnauthorized)
+	}
+
+	err := services.InsertUserToSession(c, user)
+
+	if err != nil {
```
Comment (Bokoness) [inline server/middleware/superuser_middleware.go:20]: This one should be internal error, not unauthorized

File: server/models/user.go [hunk 16]
```diff
@@ -1,14 +1,17 @@
...
 
 type User struct {
 	gorm.Model
-	Email    string `json:"email"`
-	Password string `json:"password"`
-	FullName string `json:"fullName" gorm:"column: full_name"`
+	Email    	string 	  `json:"email"`
+	Password    string    `json:"password"`
+	FullName    string    `json:"fullName" gorm:"column: full_name"`
+	IsSuperuser bool      `json:"isSuperuser" gorm:""column: is_superuser; default: false"`
```
Comment (Bokoness) [inline server/models/user.go:14]: You have a syntax error near `gorm:""`

File: server/services/auth_service.go [hunk 17]
```diff
@@ -100,3 +103,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
...
+	sess, err := GetStore(c)
+	if err != nil {
+		log.Error(fmt.Printf("this is the error %s", err))
```
Comment (Bokoness) [inline server/services/auth_service.go:109]: Try to use `log.ErrorF` instead

File: server/services/auth_service.go [hunk 18]
```diff
@@ -100,3 +103,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
...
+		return fiber.NewError(fiber.StatusInternalServerError)
```
Comment (Bokoness) [inline server/services/auth_service.go:110]: Just return the error, not the status code, the service don't know anything about response or status code, only the controller

**Review by Bokoness — CHANGES_REQUESTED** [review]

File: server/models/user.go (hunk 16 above)
Comment (Bokoness) [inline server/models/user.go:14]:
```suggestion
	IsSuperuser bool      `json:"isSuperuser" gorm:"column:is_superuser;default: false"`
```

File: server/services/auth_service.go [hunk 19]
```diff
@@ -100,3 +103,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
...
+	}
+
+	sess.Set("user", user)
+
+	if err = sess.Save(); err != nil {
+		log.Error(fmt.Printf("this is the error %s", err))
+		return fiber.NewError(fiber.StatusInternalServerError)
```
Comment (Bokoness) [inline server/services/auth_service.go:117]:
```suggestion
		return err
```

File: server/services/auth_service.go [hunk 20]
```diff
@@ -100,3 +103,29 @@ func GetUserFromSession(c *fiber.Ctx) (*models.User, error) {
...
+	}
+
+	return nil
+}
+func GetUserFromCookiesByJWT(c *fiber.Ctx) (*models.User, error) {
+	cookie := c.Cookies("user")
+
+	if cookie == "" {
+		return nil, fiber.NewError(fiber.StatusUnauthorized)
```
Comment (Bokoness) [inline server/services/auth_service.go:126]:
```suggestion
		return nil, err
```

File: server/services/auth_service.go (hunk 14 above)
Comment (Bokoness) [reply server/services/auth_service.go:107]: You can use GitHub to automatically apply my suggestion
 Anyway just add another line

File: server/services/auth_service.go (hunk 20 above)
Comment (Bokoness) [reply server/services/auth_service.go:128]: Return a custom error then, just not the status code. 
return "user not in session"

**Review by Bokoness — CHANGES_REQUESTED** [review]

**Review by Bokoness — APPROVED** [review]

**Pull request merged**